src/Core/Vault/Services/Implementations/CipherService.cs

```diff
@@ -648,10 +648,18 @@ public async Task DeleteFolderAsync(Folder folder)
             }
         }
 
-        // Init. ids for folders
+        var userfoldersIds = (await _folderRepository.GetManyByUserIdAsync(userId ?? Guid.Empty)).Select(f => f.Id).ToList();
+
+        //Assign id to the ones that don't exist in DB
+        //Need to keep the list order to create the relationships
+        List<Folder> newFolders = new List<Folder>();
```

Comment:
`folderRelationships` has a `KeyValuePair` that matches the `ciphers` and `folders` list positions. 
Below the foreach that sets the `Folder.Id ` on new folders we have another foreach on `folderRelationships` that relates the ciphers with the folders based on their positions (this is why we are keeping the `folders` list positions unchanged.

`// Create the folder associations based on the newly created folder ids
        foreach (var relationship in folderRelationships)
        {
            var cipher = ciphers.ElementAtOrDefault(relationship.Key);
            var folder = folders.ElementAtOrDefault(relationship.Value);

            if (cipher == null || folder == null)
            {
                continue;
            }

            cipher.Folders = $"{{\"{cipher.UserId.ToString().ToUpperInvariant()}\":" +
                $"\"{folder.Id.ToString().ToUpperInvariant()}\"}}";
        }`
        
        We are sending the `newFolders` to repository so that they are created.
        
        
        On Collections, there is a `CollectionCiphersTable` and we are also using the list positions to match ciphers and collections.
        
We are just sending to repository the data that needs to be created